### DISCUSSION TITLE
Call useGLTF conditionally

### DISCUSSION THREAD

**Discussion by mordv** [pmndrs/drei#569] [Q&A] [answered]
How do I use `useGLTF` if I can't guarantee there's always be an URL for the model? I can't call the hook conditionally. 
Maybe it would be better to change `path` type to `... | undefined` and handle this logic internally?

**Comment by drcmda** [comment, marked as answer]
this applies to all hooks in react, default hooks and custom. all hooks are unconditional and the solution is always composition:

```jsx
function Model({ url }) {
  const { scene } = useGLTF(url)
  return <primitive object={url} />
}

function App() {
  const url = ...
  return (
    <>
      {url && <Model url={url} />}
```

or you just let it crash and put a silent error boundary around it.

the problem with allowing undefined imo is that it breaks confidentiality. with suspense the data is guaranteed or else it goes into an error boundary. without this we'd be back to checking results. this would then seriously mess with typescript, gltfjsx, react-three-fiber/useLoader.

imo hacking conditions into hooks is not worth it, i tried in the past, this just puts the burden onto the next hook in line:

```jsx
  const { scene, animations } = useGLTF(condition ? url : undefined) // would have to return {} or deconstruct crashes
  const { actions } = useAnimations(animations) // game over ☠️
```

**Reply by liorGameDev** [reply]
Is this answer should work in 3.8.4?
When trying to use useGLTF(undefined) it seems to throw an error
@drcmda
